tls/extensions/s2n_client_psk.c

```diff
@@ -210,53 +208,20 @@ static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn,
 
-    uint16_t identities_count = 0;
-    GUARD_RESULT(s2n_count_psk_identities(wire_identities_in, &identities_count));
-    ENSURE_GT(identities_count, 0);
-    ENSURE_LTE(identities_count, MAX_NUM_OF_PSK_IDENTITIES);
+    struct s2n_offered_psk_list identity_list = { .wire_data = *wire_identities_in };
 
-    DEFER_CLEANUP(struct s2n_blob wire_identities_blob = { 0 }, s2n_free);
-    GUARD_AS_RESULT(s2n_alloc(&wire_identities_blob, identities_count * sizeof(struct s2n_psk_identity)));
-    struct s2n_psk_identity *wire_identities = (struct s2n_psk_identity*)(void*) wire_identities_blob.data;
-
-    uint16_t wire_index = 0;
-    while (s2n_stuffer_data_available(wire_identities_in) > 0) {
-        uint16_t identity_size = 0;
-        GUARD_AS_RESULT(s2n_stuffer_read_uint16(wire_identities_in, &identity_size));
-        ENSURE_GT(identity_size, 0);
-
-        uint8_t *identity_data = s2n_stuffer_raw_read(wire_identities_in, identity_size);
-        ENSURE_REF(identity_data);
-
-        wire_identities[wire_index].data = identity_data;
-        wire_identities[wire_index].length = identity_size;
-
-        /**
-         *= https://tools.ietf.org/rfc/rfc8446#section-4.2.11
-         *# For identities established externally, an obfuscated_ticket_age of 0 SHOULD be
-         *# used, and servers MUST ignore the value.
-         */
-        uint32_t obfuscated_ticket_age = 0;
-        GUARD_AS_RESULT(s2n_stuffer_read_uint32(wire_identities_in, &obfuscated_ticket_age));
+    if (conn->config->psk_selection_cb) {
+        GUARD_AS_RESULT(conn->config->psk_selection_cb(conn, &identity_list, &conn->psk_params.chosen_psk_wire_index));
 
-        wire_index++;
-    }
+        struct s2n_offered_psk wire_identity = { 0 };
+        GUARD_RESULT(s2n_offered_psk_list_get_index(&identity_list, conn->psk_params.chosen_psk_wire_index,
+                &wire_identity));
 
-    if (conn->config->psk_selection_cb) {
-        GUARD_AS_RESULT(conn->config->psk_selection_cb(conn, wire_identities, identities_count,
-                                           &conn->psk_params.chosen_psk_wire_index));
-        struct s2n_blob chosen_wire_identity = { 0 };
-        GUARD_AS_RESULT(s2n_blob_init(&chosen_wire_identity,
-                                      wire_identities[conn->psk_params.chosen_psk_wire_index].data,
-                                      wire_identities[conn->psk_params.chosen_psk_wire_index].length));
-        GUARD_RESULT(s2n_match_psk_identity(&conn->psk_params.psk_list, &chosen_wire_identity, &conn->psk_params.chosen_psk));
+        GUARD_RESULT(s2n_match_psk_identity(&conn->psk_params.psk_list, &wire_identity.identity, &conn->psk_params.chosen_psk));
     } else {
-        GUARD_RESULT(s2n_select_psk_identity(conn, wire_identities, identities_count));
+        GUARD_RESULT(s2n_select_psk_identity(conn, &identity_list));
     }
-
-    ENSURE_LT(conn->psk_params.chosen_psk_wire_index, identities_count);
     ENSURE_REF(conn->psk_params.chosen_psk);
```

Comment:
As before, do we only want success if a key is actually selected or should we support fall-back to a non-PSK handshake?

Reply:
See https://github.com/awslabs/s2n/pull/2558#discussion_r574157651